075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,368 @@
+The operator doesn't have any Kubernetes watch on the creation or update of such a resource.
+It can be only referenced by a `KafkaRebalance` custom resource so its creation doesn't trigger any action by the operator.
+The same happens when the `KafkaRebalanceSettings` custom resource is updated: no action is triggered for the operator, its updated content will be used only when a rebalancing is requested via a `KafkaRebalance` custom resource referencing those settings.
+
+### Extending the `Kafka.status` field with auto-rebalancing related state
+
+The auto-rebalancing, together with the scaling action, can take several reconciliations to finish.
+For this reason, taking into account cluster operator crashes as well, there is the need to save some state about it.
+The `Kafka` custom resource is extended by adding a new `status.cruiseControl.autoRebalance` field.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # ...
+status:
+  cruiseControl:
+    autoRebalance:
+      state: # if a scaling operation is running or what's the state of the current rebalancing
+      add-brokers: # list with brokers' IDs to add (as part of the scaling up)
+      remove-brokers: # list with brokers' IDs to remove (as part of the scaling down)
+```
+
+The `state` field provides the current state of the overall operation: if scaling is going on or what's the phase of a rebalancing operation.
+When the rebalancing is running, its value is got from the status of the corresponding `KafkaRebalance` custom resource that was created for the purpose.
+The `add-brokers` and `remove-brokers` track the brokers' IDs that are part of the scaling (up/down) actions to be useful across reconciliations, while scaling or the effective rebalancing is going on.
+Their values are got from the new `KafkaCluster.addedNodes` method (needs to be implemented) and the already existing `KafkaCluster.removedNodes` one.
+
+### Auto-rebalancing execution
+
+Before describing the process in more details, it's worth mentioning how scaling up and down happens today within the Cluster Operator reconciliation flow.
+
+The scaling up is triggered asynchronously during a reconciliation which isn't blocked on waiting for the newly added nodes to be ready.
+In the `KafkaReconcile.reconcile()` method, it is triggered via the `podSet()` method which updates the involved `StrimziPodSet`(s) so that the corresponding controller will take care of spinning up the newly added nodes while the current reconciliation just ends.
+It means that, after the scaling up is triggered, the new pods will take time to be up, running and ready, so it's likely that the auto-rebalancing will happen on a future reconciliation.
```

Comment:
This is wrong from the perspective of `KafkaReconcile.reconcile()`. But the Kafka roller, `podsReady()`, etc. will synchronize it. So from `KafkaReconcile.reconcile()` it is synchronous. The only async thing here is the Cruise Control update -> does it matter that CC doesn't know about the new nodes yet? 🤔

Reply:
>  This is wrong from the perspective of KafkaReconcile.reconcile(). But the Kafka roller, podsReady(), etc. will synchronize it. So from KafkaReconcile.reconcile() it is synchronous.

Not sure which part of the sentence you are referring to as "wrong from the perspective of ...".

> The only async thing here is the Cruise Control update -> does it matter that CC doesn't know about the new nodes yet?

I think it's explained at point `3` of the scaling up detailed process (around line 199).